server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java

```diff
@@ -0,0 +1,362 @@
+        }
+
+        @Override
+        protected void sendApplyCommit(DiscoveryNode destination, ApplyCommitRequest applyCommit,
+                                       ActionListener<TransportResponse.Empty> responseActionListener) {
+            if (this.applyCommit == null) {
+                this.applyCommit = applyCommit;
+            } else {
+                assertSame(applyCommit, this.applyCommit);
+            }
+            assertNull(pendingCommits.put(destination, responseActionListener));
+        }
+    }
+
+    DiscoveryNode n1 = CoordinationStateTests.createNode("node1");
+    DiscoveryNode n2 = CoordinationStateTests.createNode("node2");
+    DiscoveryNode n3 = CoordinationStateTests.createNode("node3");
+    Set<DiscoveryNode> discoNodes = Sets.newHashSet(n1, n2, n3);
+
+    MockNode node1 = new MockNode(Settings.EMPTY, n1);
+    MockNode node2 = new MockNode(Settings.EMPTY, n2);
+    MockNode node3 = new MockNode(Settings.EMPTY, n3);
+    List<MockNode> nodes = Arrays.asList(node1, node2, node3);
+
+    Function<DiscoveryNode, MockNode> nodeResolver = dn -> nodes.stream().filter(mn -> mn.localNode.equals(dn)).findFirst().get();
+
+    private void initializeCluster(VotingConfiguration initialConfig) {
+        node1.coordinationState.setInitialState(CoordinationStateTests.clusterState(0L, 1L, n1, initialConfig, initialConfig, 0L));
+        StartJoinRequest startJoinRequest = new StartJoinRequest(n1, 1L);
+        node1.coordinationState.handleJoin(node1.coordinationState.handleStartJoin(startJoinRequest));
+        node1.coordinationState.handleJoin(node2.coordinationState.handleStartJoin(startJoinRequest));
+        node1.coordinationState.handleJoin(node3.coordinationState.handleStartJoin(startJoinRequest));
+        assertTrue(node1.coordinationState.electionWon());
+    }
+
+    public void testSimpleClusterStatePublishing() throws InterruptedException {
+        VotingConfiguration singleNodeConfig = new VotingConfiguration(Sets.newHashSet(n1.getId()));
+        initializeCluster(singleNodeConfig);
+
+        AssertingAckListener ackListener = new AssertingAckListener(nodes.size());
+        DiscoveryNodes discoveryNodes = DiscoveryNodes.builder().add(n1).add(n2).add(n3).localNodeId(n1.getId()).build();
+        MockPublication publication = node1.publish(CoordinationStateTests.clusterState(1L, 2L,
+            discoveryNodes, singleNodeConfig, singleNodeConfig, 42L), ackListener, Collections.emptySet());
+
+        assertThat(publication.pendingPublications.keySet(), equalTo(discoNodes));
+        assertTrue(publication.pendingCommits.isEmpty());
+        publication.pendingPublications.entrySet().stream().collect(shuffle()).forEach(e -> {
+            PublishResponse publishResponse = nodeResolver.apply(e.getKey()).coordinationState.handlePublishRequest(
+                publication.publishRequest);
+            e.getValue().onResponse(new PublishWithJoinResponse(publishResponse, Optional.empty()));
```

Comment:
Could this assert that `publication.applyCommit` is null iff `n1` hasn't been processed?

Reply:
sure, fixed in f09659ceda6